src/app/clusters/network-commissioning/network-commissioning.cpp

```diff
@@ -120,25 +115,22 @@ EmberAfNetworkCommissioningError OnAddThreadNetworkCommandCallbackInternal(app::
 #if CHIP_DEVICE_CONFIG_ENABLE_THREAD
     EmberAfNetworkCommissioningError err = EMBER_ZCL_NETWORK_COMMISSIONING_ERROR_BOUNDS_EXCEEDED;
 
+    VerifyOrExit(operationalDatasetLen <= sizeof(kMaxThreadDatasetLen), );
+
     for (size_t i = 0; i < kMaxNetworks; i++)
     {
         if (sNetworks[i].mNetworkType == NetworkType::kUndefined)
         {
-            VerifyOrExit(operationalDatasetLen <= sizeof(ThreadNetworkInfo::mDataset),
-                         err = EMBER_ZCL_NETWORK_COMMISSIONING_ERROR_OUT_OF_RANGE);
-            memcpy(sNetworks[i].mData.mThread.mDataset, operationalDataset, operationalDatasetLen);
+            Thread::OperationalDataset & dataset = sNetworks[i].mData.mThread;
 
-            using ThreadDatasetLenType = decltype(sNetworks[i].mData.mThread.mDatasetLen);
-            VerifyOrExit(chip::CanCastTo<ThreadDatasetLenType>(operationalDatasetLen),
+            VerifyOrExit(dataset.Init(operationalDataset, static_cast<uint8_t>(operationalDatasetLen)) == CHIP_NO_ERROR,
                          err = EMBER_ZCL_NETWORK_COMMISSIONING_ERROR_OUT_OF_RANGE);
-            sNetworks[i].mData.mThread.mDatasetLen = static_cast<ThreadDatasetLenType>(operationalDatasetLen);
 
-            // A 64bit id for thread networks, currently, we are missing some API for getting the thread network extpanid from the
-            // dataset tlv.
-            static_assert(sizeof(kTemporaryThreadNetworkId) <= sizeof(sNetworks[i].mNetworkID),
-                          "TemporaryThreadNetworkId too long");
-            memcpy(sNetworks[i].mNetworkID, kTemporaryThreadNetworkId, sizeof(kTemporaryThreadNetworkId));
-            sNetworks[i].mNetworkIDLen = sizeof(kTemporaryThreadNetworkId);
+            static_assert(sizeof(sNetworks[i].mNetworkID) >= Thread::kSizeExtendedPanId,
+                          "Network ID must be larger than Thread extended PAN ID!");
+            SuccessOrExit(
+                dataset.GetExtendedPanId(reinterpret_cast<uint8_t(&)[Thread::kSizeExtendedPanId]>(sNetworks[i].mNetworkID)));
```

Comment:
Can you cite a reason why this cast wouldn't be undefined behavior? e.g. see https://en.cppreference.com/w/cpp/language/reinterpret_cast for a summary of defined uses.

Reply:
I think you are right. I added a temporary variable to avoid this kind of casts.